### PR TITLE
Accessor improvements

### PR DESCRIPTION
**Pull request by thoasm**
This PR extends the the accessor by adding overloads to math functions for the accessor references, so `conj()` can be called directly on an accessor with a custom reference. Tests for both the math functions and for all functionalities for the references was also added.

Additionally, the functions `as_cuda_accessor` and `as_hip_accessor` were added to help with the conversion `std::complex` -> `thrust::complex`.

TODO:
- [ ] Merge #677
- [ ] Merge #719
- [ ] Adopt merged code to this PR by removing work-arounds

### CODE REVIEW

**Comment by thoasm** [comment]
format!

File: core/test/accessor/scaled_reduced_row_major_reference.cpp [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+    auto get_ref() { return ref_type{&storage, scalar}; }
+    auto get_const_ref() { return const_ref_type{&storage, scalar}; }
+
+    auto get_conv_storage() { return scalar * static_cast<ar_type>(storage); }
+
+    ar_type scalar;
+    st_type storage;
+};
+
+// using ReferenceTypes = ::testing::Types<std::tuple<char, short>>;
```
Comment (yhmtsai) [inline core/test/accessor/scaled_reduced_row_major_reference.cpp:77]:
```suggestion
```

File: include/ginkgo/core/base/math.hpp [hunk 2]
```diff
@@ -817,53 +818,162 @@ GKO_INLINE GKO_ATTRIBUTES constexpr T min(const T& x, const T& y)
...
- * @tparam T  type of the object
- *
- * @param x  the object
- *
- * @return real part of the object (by default, the object itself)
+ * @internal
+ * Tests if a member function `Ref::to_arithmetic_type` exists
+ */
+template <typename Ref, typename Dummy = xstd::void_t<>>
+struct has_to_arithmetic_type : std::false_type {
```
Comment (yhmtsai) [inline include/ginkgo/core/base/math.hpp:829]: Is this part the same as accessor one?

**Review by yhmtsai — COMMENTED** [review]
first part of my review

File: accessor/reference_helper.hpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+template <typename Reference, typename ArithmeticType>
+struct enable_reference_operators {
+    using arithmetic_type = std::remove_cv_t<ArithmeticType>;
+
+    constexpr GKO_ACC_ATTRIBUTES GKO_ACC_INLINE arithmetic_type
+    to_arithmetic_type() const
+    {
+        return static_cast<const Reference*>(this)
+            ->Reference::operator arithmetic_type();
+    }
```
[the remark this answers is not in the document]
Comment (thoasm) [reply accessor/reference_helper.hpp:92]: This function calls the cast operator (`operator arithmetic_type()`) of the `this` object.
Because this cast operator is defined in the `Reference` class, It needs to be cast to `Reference` first, before the operator is called. Since this is a Mixin and `Reference` is supposed to inherit from it, this cast should always be succesfull (also, I assume if it would not be inherited, this should yield in a compiler error).

In summary, this has 2 layers:
1. This calls the cast operator to `arithmetic_type` that must be present (otherwise, a proxy reference would not make sense). If it is not present, it will result in a compilation error.
2. Since `enable_reference_operators` does not have that cast operator, it needs to be cast to `Reference` first before the operator can be found.

File: accessor/utils.hpp [hunk 4]
```diff
@@ -166,6 +166,81 @@ struct are_all_integral<First, Args...>
...
+ *    result of that function
+ * 2. Otherwise, if the type `Ref::arithmetic_type` exists, it will return the
+ *    result of `static_cast<Ref::arithmetic_type>(ref)`
+ * 3. Otherwise, it will return `ref` itself.
+ */
+template <typename Ref>
+constexpr GKO_ACC_ATTRIBUTES
+    std::enable_if_t<has_to_arithmetic_type<Ref>::value,
+                     typename has_to_arithmetic_type<Ref>::type>
+    to_arithmetic_type(const Ref& ref)
```
[the remark this answers is not in the document]
Comment (thoasm) [reply accessor/utils.hpp:219]: This actually does the conversion as well, and not just return the type.

File: core/test/accessor/reduced_row_major_reference.cpp [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+    const ar_type expected_self_res = static_cast<ar_type>(
+        this->get_conv_storage() * this->get_conv_storage());
+
+    auto res1 = mult * this->get_ref();
+    auto res2 = this->get_ref() * mult;
+    auto res3 = this->get_const_ref() * mult;
+    auto res4 = mult * this->get_const_ref();
+    // Not properly supported:
+    // auto self_res1 = this->get_ref() * this->get_const_ref();
+    // auto self_res2 = this->get_const_ref() * this->get_ref();
```
[the remark this answers is not in the document]
Comment (thoasm) [reply core/test/accessor/reduced_row_major_reference.cpp:134]: Compilation error, as the operator choice is ambiguous (there will be 2 operators that match equally well, so the compiler can't choose which to take).

File: accessor/cuda_helper.hpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+/**
+ * Changes the types and reinterprets the passed in range pointers as a CUDA
+ * types.
+ *
+ * @param r  the range which pointers need to be reinterpreted
+ *
+ * @return `r` with appropriate types and reinterpreted to CUDA pointers
+ */
+template <int dim, typename Type1, typename Type2>
+GKO_ACC_INLINE auto as_cuda_range(
```
[the remark this answers is not in the document]
Comment (thoasm) [reply accessor/cuda_helper.hpp:143]: I prefer to leave it as it is, since this way, it is more descriptive if you use it in a non-Ginkgo environment. I will look if it is possible to use `as_cuda_range` inside `as_cuda_type`, so it is transparent inside Ginkgo.

File: accessor/accessor_helper.hpp [hunk 7]
```diff
@@ -59,49 +59,47 @@ namespace detail {
...
 
-    template <typename FirstType, typename... Indices>
-    static constexpr GKO_ACC_ATTRIBUTES ValueType
-    compute(const std::array<DimensionType, total_dim>& size,
-            const std::array<ValueType, (total_dim > 1 ? total_dim - 1 : 0)>&
-                stride,
-            FirstType first, Indices&&... idxs)
+    template <typename SizeType, typename... Indices>
+    static constexpr GKO_ACC_ATTRIBUTES IndexType compute(
+        const std::array<SizeType, total_dim>& size,
```
[the remark this answers is not in the document]
Comment (thoasm) [reply accessor/accessor_helper.hpp:76]: Actually, `IndexType` and `SizeType` are currently different types (in a different version, they might be unified). `IndexType` is the type for the indices, and `SizeType` is the type of the dimension and stride.
Inside Accessors, you (somewhat need) an `std::array` containing the max sizes (basically the non-Ginkgo alternative to `gko::dim<>`) and for most of them, another `std::array` containing the strides.
Currently, both the dim and the strides use `size_type` internally, which is different of `IndexType`.
The reason why I made `IndexType` explicit (part of the struct and not a template parameter) was to:
1. Force the input to actually be an integer type and not an arbitrary type
2. Unify the index computation. Before, it was possible to have 32bit and 64bit inputs mixed, which performed quite badly on some architectures.

However, it might make sense to immediately move to `int64_t` being the `size_type` inside the accessor since it is not a public interface. I will look what we need to do in order to achieve that.

**Review by tcojean — APPROVED** [review]
LGTM!
I think it only needs some rebasing?

**Comment by thoasm** [comment]
format!
